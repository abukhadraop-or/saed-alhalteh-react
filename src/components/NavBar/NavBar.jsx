import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import {
  Container,
  LinksList,
  Logo,
  MenuBlock,
  MenuButton,
  MenuIcon,
  NavItem,
} from './nav-bar.styles';

/**
 * {String} The logo to be shown in the Nav-bar.
 */
const logo = 'Conduit';

/**
 * {Object[]} that contains the Links title and url to be shown in the Nav-bar.
 */
const items = [
  {
    label: 'Home',
    to: '/',
  },
  {
    label: 'Sign In',
    to: '/signIn',
  },
  {
    label: 'Sign Up',
    to: '/signUp',
  },
];

/**
 * Component showing the main navigation bar of the application.
 *
 * @return {JSX.Element} Navigation bar component.
 */
function NavBar() {
  const [isOpen, setIsOpen] = useState(false);

  /**
   * Revert the status of the menu.
   */
  const handleMenuClick = () => {
    setIsOpen(!isOpen);
  };

  return (
    <>
      <Container>
        <Logo>
          <Link to="/">{logo}</Link>
        </Logo>
        <LinksList>
          {items.map((item) => (
            <NavItem key={item.label}>
              <Link to={item.to}>{item.label}</Link>
            </NavItem>
          ))}
        </LinksList>
        <MenuButton onClick={() => handleMenuClick()}>
          <MenuIcon />
        </MenuButton>
      </Container>
      {isOpen && (
        <MenuBlock>
          {items.map((item) => (
            <NavItem>
              <Link to={item.to} onClick={() => handleMenuClick()}>
                {item.label}
              </Link>
            </NavItem>
          ))}
        </MenuBlock>
      )}
    </>
  );
}

export default NavBar;
