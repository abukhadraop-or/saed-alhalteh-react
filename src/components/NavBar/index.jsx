import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import {
  Container,
  LinksList,
  Logo,
  MenuBlock,
  MenuButton,
  MenuIcon,
  NavItem,
} from './nav-bar.styles';

/**
 * @type {string} The logo.
 */
const logo = 'Conduit';

/**
 * @type {object<string, string>[]} Links label and url.
 */
const items = [
  {
    label: 'Home',
    to: '/',
  },
  {
    label: 'Sign In',
    to: '/signIn',
  },
  {
    label: 'Sign Up',
    to: '/signUp',
  },
];

const menuBlock = React.createRef();

/**
 * The main navigation bar of the application.
 *
 * @return {JSX.Element} NavBar component.
 */
function NavBar() {
  const [isOpen, setIsOpen] = useState(false);

  /**
   * Handle menuButton on blur event.
   *
   * @param {object} event On blur event.
   */
  const handleBlur = (event) => {
    const target = event && event.relatedTarget;

    if (!target || target.parentElement.parentElement !== menuBlock.current) {
      setIsOpen(false);
    }
  };

  return (
    <>
      <Container>
        <Logo>
          <Link to="/">{logo}</Link>
        </Logo>
        <LinksList>
          {items.map((item) => (
            <NavItem key={item.label}>
              <Link to={item.to}>{item.label}</Link>
            </NavItem>
          ))}
        </LinksList>
        <MenuButton onClick={() => setIsOpen(!isOpen)} onBlur={(event) => handleBlur(event)}>
          <MenuIcon />
        </MenuButton>
      </Container>
      {isOpen && (
        <MenuBlock ref={menuBlock}>
          {items.map((item) => (
            <NavItem key={item.label}>
              <Link to={item.to} onClick={() => setIsOpen(!isOpen)}>
                {item.label}
              </Link>
            </NavItem>
          ))}
        </MenuBlock>
      )}
    </>
  );
}

export default NavBar;
