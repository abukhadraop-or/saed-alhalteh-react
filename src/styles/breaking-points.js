const breakingPoints = {
  mobile: '30rem',
  tablets: '48rem',
  smallScreen: '64.0625rem',
  largeScreen: '75rem',
};

export default breakingPoints;
