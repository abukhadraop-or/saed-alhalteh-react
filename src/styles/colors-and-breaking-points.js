export const colors = {
  black: '#000',
  blueViolet: '#4F52B0',
  concrete: '#f3f3f3',
  dustyGrey: '#999',
  edward: '#acb9b4',
  fern: '#5cb85c',
  goblin: '#347634',
  gray: '#868686',
  mineShaft: '#333333',
  oBlack: 'rgba(0,0,0,0.75)',
  o3Black: 'rgba(0,0,0,0.3)',
  oFern: 'rgba(92,184,92,0.5)',
  silver: '#c8c8c8',
  white: '#fff',
};

export const breakingPoints = {
  mobile: '480px',
  tablets: '768px',
  smallScreen: '1025px',
  largeScreen: '1200px',
};
