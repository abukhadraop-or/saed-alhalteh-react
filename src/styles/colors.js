const colors = {
  black: '#000000',
  blueViolet: '#4f52b0',
  concrete: '#f3f3f3',
  darkGreen: '#347634',
  dustyGrey: '#999999',
  gray: '#868686',
  green: '#5cb85c',
  lightGrey: '#acb9b4',
  mineShaft: '#333333',
  partiallyDarkBlack: 'rgba(0,0,0,0.75)',
  silver: '#c8c8c8',
  transparentBlack: 'rgba(0,0,0,0.3)',
  transparentGreen: 'rgba(92,184,92,0.5)',
  white: '#ffffff',
};

export default colors;
