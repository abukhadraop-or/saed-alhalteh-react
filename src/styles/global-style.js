import { createGlobalStyle } from 'styled-components';

const GlobalStyle = createGlobalStyle`
  body {
    font-family:'Titillium Web', sans-serif;
    margin: 0;
  }
`;

export default GlobalStyle;
