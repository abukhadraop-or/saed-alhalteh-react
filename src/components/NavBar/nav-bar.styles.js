import styled from 'styled-components';
import { ReactComponent as HamburgerMenu } from 'assets/Icons/hamburgerMenu.svg';
import { colors, breakingPoints } from 'shared/styles';

export const Container = styled.div`
  display: flex;
  justify-content: space-between;
  padding: 1.25rem;

  & a {
    color: ${colors.gray};
    padding: 0.625rem;
    text-decoration: none;

    &:hover {
      color: black;
    }
  }
`;

export const Logo = styled.div`
  & a {
    color: ${colors.fern};
    font: 600 1.5rem 'Titillium Web', sans-serif;

    &:hover {
      color: ${colors.fern};
    }
  }
`;

export const LinksList = styled.ul`
  align-self: center;
  margin: 0;
  padding: 0;
  list-style-type: none;

  @media (max-width: ${breakingPoints.tablets}) {
    display: none;
  }
`;

export const NavItem = styled.li`
  display: inline-block;
`;

export const MenuIcon = styled(HamburgerMenu)`
  align-self: center;
  cursor: pointer;
  display: none;
  fill: ${colors.fern};

  @media (max-width: ${breakingPoints.tablets}) {
    display: inline-block;
  }
`;

export const MenuBlock = styled.div`
  background-color: ${colors.oBlack};
  border: 0.0625rem solid ${colors.black};
  border-radius: 0.9375rem;
  min-height: 11.875rem;
  position: absolute;
  right: 1.25rem;
  top: 3.125rem;
  width: 7.5rem;

  & a {
    color: ${colors.white};
    display: block;
    font-weight: 600;
    padding: 1.25rem;
    text-decoration: none;
  }

  & li {
    display: block;
    margin: 1.25rem 0;

    & :hover {
      background-color: ${colors.blueViolet};
    }
  }
`;
