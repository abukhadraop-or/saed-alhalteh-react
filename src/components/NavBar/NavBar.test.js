import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import Enzyme, { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import NavBar from '.';
import { MenuBlock, MenuButton } from './nav-bar.styles';

Enzyme.configure({ adapter: new Adapter() });

describe('<NavBar /> tests', () => {
  let wrapper;
  const elements = {
    MenuButton: () => wrapper.find(MenuButton),
    MenuBlock: () => wrapper.find(MenuBlock),
  };

  beforeEach(() => {
    wrapper = shallow(<NavBar />);
  });

  it('matches the snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('toggles the displays of the MenuBlock when MenuButton is clicked', () => {
    elements.MenuButton().simulate('click');

    expect(elements.MenuBlock().length).toBe(1);

    elements.MenuButton().simulate('click');

    expect(elements.MenuBlock().length).toBe(0);
  });

  it('hides the MenuBlock when a link is clicked', () => {
    elements.MenuButton().simulate('click');
    elements.MenuBlock().find('Link').at(0).simulate('click');

    expect(elements.MenuBlock().length).toBe(0);
  });
});
