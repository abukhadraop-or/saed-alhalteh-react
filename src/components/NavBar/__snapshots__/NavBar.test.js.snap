// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavBar /> tests matches the snapshot 1`] = `
<Fragment>
  <styled.div>
    <styled.div>
      <Link
        to="/"
      >
        Conduit
      </Link>
    </styled.div>
    <styled.ul>
      <styled.li
        key="Home"
      >
        <Link
          to="/"
        >
          Home
        </Link>
      </styled.li>
      <styled.li
        key="Sign In"
      >
        <Link
          to="/signIn"
        >
          Sign In
        </Link>
      </styled.li>
      <styled.li
        key="Sign Up"
      >
        <Link
          to="/signUp"
        >
          Sign Up
        </Link>
      </styled.li>
    </styled.ul>
    <styled.button
      onClick={[Function]}
    >
      <Styled(Component) />
    </styled.button>
  </styled.div>
</Fragment>
`;
